src/protocols/http/hpr_roaming_protocol.erl

```diff
@@ -0,0 +1,541 @@
+            DataRate = hpr_lorawan:index_to_datarate(Region, DR),
+
+            Delay1 =
+                case Delay0 of
+                    N when N < 2 -> 1;
+                    N -> N
+                end,
+
+            Timeout =
+                case DeviceClass of
+                    <<"C">> ->
+                        immediate;
+                    <<"A">> ->
+                        hpr_roaming_utils:uint32(PacketTime + (Delay1 * ?RX1_DELAY) + ?RX1_DELAY)
+                end,
+
+            DownlinkPacket = new_downlink(
+                hpr_roaming_utils:hexstring_to_binary(Payload),
+                Timeout,
+                Frequency,
+                DataRate,
+                undefined
+            ),
+
+            case hpr_roaming_downlink:lookup_handler(TransactionID) of
+                {error, _} = Err ->
+                    Err;
+                {ok, ResponseStream} ->
+                    {downlink, PayloadResponse, {ResponseStream, DownlinkPacket},
+                        {DestURL, FlowType}}
+            end
+    end.
+
+rx2_from_dlmetadata(
+    #{
+        <<"DataRate2">> := DR,
+        <<"DLFreq2">> := Frequency
+    },
+    PacketTime,
+    Region,
+    Timeout
+) ->
+    try hpr_lorawan:index_to_datarate(Region, DR) of
+        DataRate ->
+            window(
+                hpr_roaming_utils:uint32(PacketTime + Timeout),
+                Frequency,
+                DataRate
+            )
+    catch
+        Err ->
+            lager:warning("skipping rx2, bad dr_to_datar(~p, ~p) [err: ~p]", [Region, DR, Err]),
+            undefined
+    end;
+rx2_from_dlmetadata(_, _, _, _) ->
+    lager:debug("skipping rx2, no details"),
+    undefined.
+
+%% ------------------------------------------------------------------
+%% Tokens
+%% ------------------------------------------------------------------
+
+-spec make_uplink_token(transaction_id(), region(), non_neg_integer(), binary(), atom()) -> token().
+make_uplink_token(TransactionID, Region, PacketTime, DestURL, FlowType) ->
+    Parts = [
+        erlang:integer_to_binary(TransactionID),
+        erlang:atom_to_binary(Region),
+        erlang:integer_to_binary(PacketTime),
+        DestURL,
+        erlang:atom_to_binary(FlowType)
+    ],
+    Token0 = lists:join(?TOKEN_SEP, Parts),
+    Token1 = erlang:iolist_to_binary(Token0),
+    hpr_roaming_utils:binary_to_hexstring(Token1).
+
+-spec parse_uplink_token(token()) ->
+    {ok, transaction_id(), region(), non_neg_integer(), dest_url(), flow_type()} | {error, any()}.
+parse_uplink_token(<<"0x", Token/binary>>) ->
+    parse_uplink_token(Token);
+parse_uplink_token(Token) ->
+    Bin = hpr_roaming_utils:hex_to_binary(Token),
+    case binary:split(Bin, ?TOKEN_SEP, [global]) of
+        [TransactionIDBin, RegionBin, PacketTimeBin, DestURLBin, FlowTypeBin] ->
+            TransactionID = erlang:binary_to_integer(TransactionIDBin),
+            Region = erlang:binary_to_atom(RegionBin),
```

Comment:
Should the `Region` atom already exist?

Reply:
Are you suggesting using erlang:binary_to_existing_atom/2 ?

Reply:
I made the change.